### DISCUSSION TITLE
Is there an easier way to cast away an atomic specifier for subsequent shaders/dispatches?

### DISCUSSION THREAD

**Discussion by kjpgit** [gpuweb/gpuweb#4438] [Q&A] [answered]
I'm currently building a hashtable using atomic<i32> with one shader, and then subsequent dispatches/shaders only read from that hashtable.   Currently I'm doing this (copying and pasting a struct, just removing the atomic), and it feels bad:

    struct MiscData {
        // This is calculated during rough shape processing, when appending to fine buffer
        num_fine_shapes: atomic<i32>,
    
        // This allocates memory during bin step 2
        num_fine_pointers: atomic<i32>,
    
        // This is calculated during binning step 1, and can be read by the GUI
        @align(4096) num_shapes_per_tile: array<array<atomic<i32>, NUM_TILES_X>, NUM_TILES_Y>
    };
    
    
    struct MiscDataRead {
        num_fine_shapes: i32,
        num_fine_pointers: i32,
        @align(4096) num_shapes_per_tile: array<array<i32, NUM_TILES_X>, NUM_TILES_Y>,
    };

**Comment by alan-baker** [comment, marked as answer]
Unfortunately, not currently. If you could restructure your code generation you might be able to make it a bit easier on yourself if you use a type alias:
```
alias DATA_TYPE = atomic<i32>; // or i32 in the other case

struct MiscData {
  num_fine_shapes : DATA_TYPE
  num_fine_pointers : DATA_TYPE,
  @align(4096) num_shapes_per_tile : array<array<DATA_TYPE, NUM_TILES_X>, NUM_TILES_Y>
}
```

That might save you a little effort, but doesn't fully address the problem.

**Reply by stefnotch** [reply]
Would it be reasonable to allow for a "readonly" usage of an atomic? As in, allowing for
```wgsl
// This is currently not possible, since it leads to an error
@group(0) @binding(0) var<storage, read> buffer : MiscData;
```

The error roughly being
```
Error while parsing WGSL: :26:42 error: atomic variables in <storage> address space must have read_write access mode
@group(0) @binding(0) var<storage, read> buffer : MiscData;
                                         ^^^^^^^^^^^^^

:16:1 note: atomic sub-type of 'MiscData' is declared here
num_fine_shapes: atomic<i32>,
^^^^^
```

**Reply by alan-baker** [reply]
You could open an issue for that as a feature request. The validation on atomic built-in functions could be changed to have the access restrictions instead of the type. Only `atomicLoad` would be allowed in the read-only case. Hopefully, drivers would just optimize them to regular loads.

**Reply by stefnotch** [reply]
@alan-baker Sounds good, I opened a request https://github.com/gpuweb/gpuweb/issues/4574
